EIPS/eip-beacon_state_root_in_evm.md

```diff
@@ -0,0 +1,169 @@
+| `G_beacon_state_root`     | 20                                           | gas
+
+### Background
+
+The method of injecting the beacon state root in this EIP follows the general strategy of [EIP-4399](./eip-4399.md) to make a post-merge change to the EVM integrating information from the beacon chain. This EIP along with [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675) should be taken as relevant background to understand the particular approach of this EIP.
+
+The method for exposing the state root data via opcode is inspired by [EIP-2935](./eip-2935.md).
+
+### Block structure and validity
+
+Beginning at the execution timestamp `FORK_TIMESTAMP`, execution clients **MUST**:
+
+1. set the value of the `ommers` field in the block to an RLP list with one element: the 32 byte [hash tree root](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#merkleization) of the [beacon state](https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/beacon-chain.md#beaconstate) from the previous slot to this block.
+
+2. set the value of the `ommersHash` field in the block header to the Keccak256 hash of the `ommers` field.
+
+```python
+beaconStateRoot = <32 byte value> # provided by consensus client
+ommers = RLP([beaconStateRoot])   # in the block body
+ommersHash = Keccak256(ommers)    # in the block header
+```
+
+3. Add the block validation that the `ommersHash` does indeed match the expected committment given the `ommers` value.
+
+### EVM changes
+
+#### Block processing
+
+At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions), write the beacon state root provided in the block into the storage of the smart contract at `HISTORY_STORAGE_ADDRESS`. This data is keyed by the block number.
+
+In pseudocode:
+
+```python
+beacon_state_root = block.ommers[0]
+sstore(HISTORY_STORAGE_ADDRESS, block.number, beacon_state_root)
+```
+
+#### New opcode
+
+Beginning at the execution timestamp `FORK_TIMESTAMP`, introduce a new opcode `BEACON_STATE_ROOT` at `OPCODE_VALUE`. This opcode consumes one word from the stack encoding the block number for the root. The opcode has a gas cost of `G_beacon_state_root`.
+
+The result of executing this opcode leaves one word on the stack corresponding to a read of the history contract's storage; in pseudocode:
+
+```python
+block_number = evm.stack.pop()
+sload(HISTORY_STORAGE_ADDRESS, block_number)
+```
+
+If there is no root stored at the requested block number, the opcode follows the existing EVM semantics of `sload` returning `0`.
+
+###  Other considerations
+
+The [Engine API `PayloadAttributes` message](https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#payloadattributesv1) will be updated with an additional field to send the relevant beacon state root from the consensus client to the execution client when preparing a block.
```

Comment:
Recommend removing this section and leaving it up to a separate changeset against the execution API specification.  This EIP can stand on its own without reference to *how* the execution client gets the data, and in general I prefer to keep changesets as isolated as possible.

Reply:
makes sense